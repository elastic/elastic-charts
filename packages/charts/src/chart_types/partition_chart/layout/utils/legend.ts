/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { CategoryKey } from '../../../../common/category';
import { map } from '../../../../common/iterables';
import { LegendItem } from '../../../../common/legend';
import { LegendPositionConfig } from '../../../../specs/settings';
import { isHierarchicalLegend } from '../../../../utils/legend';
import { Layer } from '../../specs';
import { PartitionLayout } from '../types/config_types';
import { QuadViewModel } from '../types/viewmodel_types';

function makeKey(...keyParts: CategoryKey[]): string {
  return keyParts.join('---');
}

function compareTreePaths(
  { index: oiA, innerIndex: iiA, path: a }: QuadViewModel,
  { index: oiB, innerIndex: iiB, path: b }: QuadViewModel,
): number {
  if (oiA !== oiB) return oiA - oiB;
  if (iiA !== iiB) return iiA - iiB;
  for (let i = 0; i < Math.min(a.length, b.length); i++) {
    const diff = a[i].index - b[i].index;
    if (diff) {
      return diff;
    }
  }
  return a.length - b.length; // if one path is fully contained in the other, then parent (shorter) goes first
}

/** @internal */
export function getLegendItems(
  id: string,
  layers: Layer[],
  flatLegend: boolean | undefined,
  legendMaxDepth: number,
  legendPosition: LegendPositionConfig,
  quadViewModel: QuadViewModel[],
  partitionLayout: PartitionLayout | undefined,
): LegendItem[] {
  const uniqueNames = new Set(map(quadViewModel, ({ dataName, fillColor }) => makeKey(dataName, fillColor)));
  const useHierarchicalLegend = isHierarchicalLegend(flatLegend, legendPosition);

  const formattedLabel = ({ dataName, depth }: QuadViewModel) => {
    const formatter = layers[depth - 1]?.nodeLabel;
    return formatter ? formatter(dataName) : dataName;
  };

  function compareNames(aItem: QuadViewModel, bItem: QuadViewModel): number {
    const a = formattedLabel(aItem);
    const b = formattedLabel(bItem);
    return a < b ? -1 : a > b ? 1 : 0;
  }

  function descendingValues(aItem: QuadViewModel, bItem: QuadViewModel): number {
    return aItem.depth - bItem.depth || bItem.value - aItem.value;
  }

  const excluded: Set<string> = new Set();
  const items = quadViewModel.filter(({ depth, dataName, fillColor }) => {
    if (legendMaxDepth !== null && depth > legendMaxDepth) {
      return false;
    }
    if (!useHierarchicalLegend) {
      const key = makeKey(dataName, fillColor);
      if (uniqueNames.has(key) && excluded.has(key)) {
        return false;
      }
      excluded.add(key);
    }
    return true;
  });

  items.sort(
    partitionLayout === PartitionLayout.waffle // waffle has inherent top to bottom descending order
      ? descendingValues
      : flatLegend
      ? compareNames
      : compareTreePaths,
  );

  return items.map<LegendItem>((item) => {
    const { dataName, fillColor, depth, path } = item;
    return {
      color: fillColor,
      label: formattedLabel(item),
      childId: dataName,
      depth: useHierarchicalLegend ? depth - 1 : 0,
      path,
      seriesIdentifiers: [{ key: dataName, specId: id }],
      keys: [],
    };
  });
}
