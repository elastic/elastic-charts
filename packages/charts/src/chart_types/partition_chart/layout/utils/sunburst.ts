/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { ArrayEntry, childrenAccessor, HierarchyOfArrays } from './group_by_rollup';
import { Origin } from '../../../../common/geometry';
import { Part } from '../../../../common/text_utils';

/** @internal */
export function sunburst(
  outerNodes: HierarchyOfArrays,
  areaAccessor: (e: ArrayEntry) => number,
  { x0: outerX0, y0: outerY0 }: Origin,
  clockwiseSectors: boolean,
  specialFirstInnermostSector: boolean,
  heightStep: number = 1,
): Array<Part> {
  const result: Array<Part> = [];
  const laySubtree = (nodes: HierarchyOfArrays, { x0, y0 }: Origin, depth: number) => {
    let currentOffsetX = x0;
    const nodeCount = nodes.length;
    for (let i = 0; i < nodeCount; i++) {
      const index = clockwiseSectors ? i : nodeCount - i - 1;
      const node = nodes[depth === 1 && specialFirstInnermostSector ? (index + 1) % nodeCount : index];
      const area = areaAccessor(node);
      result.push({ node, x0: currentOffsetX, y0, x1: currentOffsetX + area, y1: y0 + heightStep });
      const children = childrenAccessor(node);
      if (children.length > 0) {
        laySubtree(children, { x0: currentOffsetX, y0: y0 + heightStep }, depth + 1);
      }
      currentOffsetX += area;
    }
  };
  laySubtree(outerNodes, { x0: outerX0, y0: outerY0 }, 0);
  return result;
}
