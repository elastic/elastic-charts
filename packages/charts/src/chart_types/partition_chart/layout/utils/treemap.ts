/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { $Values as Values } from 'utility-types';

import { ArrayEntry, CHILDREN_KEY, DEPTH_KEY, entryValue, HierarchyOfArrays } from './group_by_rollup';
import { GOLDEN_RATIO } from '../../../../common/constants';
import { Pixels } from '../../../../common/geometry';
import { Part } from '../../../../common/text_utils';

const MAX_U_PADDING_RATIO = 0.0256197; // this limits area distortion to <10% (which occurs due to pixel padding) with very small rectangles
const MAX_TOP_PADDING_RATIO = 0.33; // this limits further area distortion to ~33%

interface LayoutElement {
  nodes: HierarchyOfArrays;
  dependentSize: number;
  sectionSizes: number[];
  sectionOffsets: number[];
}

function layVector(
  nodes: HierarchyOfArrays,
  independentSize: number,
  areaAccessor: (e: ArrayEntry) => number,
): LayoutElement {
  const area = nodes.reduce((p, n) => p + areaAccessor(n), 0);
  const dependentSize = independentSize === 0 ? 0 : area / independentSize; // here we lose a bit of accuracy
  let currentOffset = 0;
  const sectionOffsets = [currentOffset];
  const sectionSizes = nodes.map((e, i) => {
    const sectionSize = dependentSize === 0 ? 0 : areaAccessor(e) / dependentSize; // here we gain back a bit of accuracy
    if (i < nodes.length - 1) sectionOffsets.push((currentOffset += sectionSize));
    return sectionSize;
  });

  return { nodes, dependentSize, sectionSizes, sectionOffsets }; // descriptor for a vector (column or row) of elements (nodes)
}

/** @internal */
export function leastSquarishAspectRatio({ sectionSizes, dependentSize }: LayoutElement) {
  if (dependentSize === 0) {
    return 1;
  }
  return sectionSizes.reduce((p, n) => Math.min(p, n / dependentSize, dependentSize / n), 1);
}

const NullLayoutElement: LayoutElement = {
  nodes: [],
  dependentSize: NaN,
  sectionSizes: [],
  sectionOffsets: [],
};

/**
 * Specifies whether partitions are laid out horizontally, vertically or treemap-like tiling for preferably squarish aspect ratios
 * @public
 */
export const LayerLayout = Object.freeze({
  horizontal: 'horizontal' as const,
  vertical: 'vertical' as const,
  squarifying: 'squarifying' as const,
});

/**
 * Specifies whether partitions are laid out horizontally, vertically or treemap-like tiling for preferably squarish aspect ratios
 * @public
 */
export type LayerLayout = Values<typeof LayerLayout>; // could use ValuesType<typeof HierarchicalChartTypes>

function bestVector(
  nodes: HierarchyOfArrays,
  height: number,
  areaAccessor: (e: ArrayEntry) => number,
  layout: LayerLayout | null,
): LayoutElement {
  let previousWorstAspectRatio = -1;
  let currentWorstAspectRatio = 0;

  let previousVectorLayout: LayoutElement = NullLayoutElement;
  let currentVectorLayout: LayoutElement = NullLayoutElement;
  let currentCount = 1;
  do {
    previousVectorLayout = currentVectorLayout;
    previousWorstAspectRatio = currentWorstAspectRatio;
    currentVectorLayout = layVector(nodes.slice(0, currentCount), height, areaAccessor);
    currentWorstAspectRatio = leastSquarishAspectRatio(currentVectorLayout);
  } while (currentCount++ < nodes.length && (layout || currentWorstAspectRatio > previousWorstAspectRatio));

  return layout || currentWorstAspectRatio >= previousWorstAspectRatio ? currentVectorLayout : previousVectorLayout;
}

function vectorNodeCoordinates(vectorLayout: LayoutElement, x0Base: number, y0Base: number, vertical: boolean) {
  const { nodes, dependentSize, sectionSizes, sectionOffsets } = vectorLayout;
  return nodes.map((e: ArrayEntry, i: number) => {
    const offset = sectionOffsets[i] ?? 0;
    const size = sectionSizes[i] ?? 0;
    const x0 = vertical ? x0Base + offset : x0Base;
    const y0 = vertical ? y0Base : y0Base + offset;
    const x1 = vertical ? x0 + size : x0 + dependentSize;
    const y1 = vertical ? y0 + dependentSize : y0 + size;
    return { node: e, x0, y0, x1, y1 };
  });
}

/** @internal */
export const getTopPadding = (requestedTopPadding: number, fullHeight: Pixels) =>
  Math.min(requestedTopPadding, fullHeight * MAX_TOP_PADDING_RATIO);

/** @internal */
export function treemap(
  nodes: HierarchyOfArrays,
  areaAccessor: (e: ArrayEntry) => number,
  topPaddingAccessor: (e: ArrayEntry) => number,
  paddingAccessor: (e: ArrayEntry) => number,
  {
    x0: outerX0,
    y0: outerY0,
    width: outerWidth,
    height: outerHeight,
  }: { x0: number; y0: number; width: number; height: number },
  layouts: LayerLayout[],
): Array<Part> {
  if (nodes.length === 0) return [];
  // some bias toward horizontal rectangles with a golden ratio of width to height
  const depth = (nodes[0]?.[1][DEPTH_KEY] ?? 1) - 1;
  const layerLayout = layouts[depth] ?? null;
  const vertical = layerLayout === LayerLayout.vertical || (!layerLayout && outerWidth / GOLDEN_RATIO <= outerHeight);
  const independentSize = vertical ? outerWidth : outerHeight;
  const vectorElements = bestVector(nodes, independentSize, areaAccessor, layerLayout);
  const vector = vectorNodeCoordinates(vectorElements, outerX0, outerY0, vertical);
  const { dependentSize } = vectorElements;
  return vector
    .concat(
      ...vector.map(({ node, x0, y0, x1, y1 }) => {
        const childrenNodes = entryValue(node)[CHILDREN_KEY];
        if (childrenNodes.length === 0) {
          return [];
        }
        const fullWidth = x1 - x0;
        const fullHeight = y1 - y0;
        const uPadding = Math.min(
          paddingAccessor(node),
          fullWidth * MAX_U_PADDING_RATIO * 2,
          fullHeight * MAX_U_PADDING_RATIO * 2,
        );
        const topPadding = getTopPadding(topPaddingAccessor(node), fullHeight);
        const width = fullWidth - 2 * uPadding;
        const height = fullHeight - uPadding - topPadding;
        return treemap(
          childrenNodes,
          (d) => ((width * height) / (fullWidth * fullHeight)) * areaAccessor(d),
          topPaddingAccessor,
          paddingAccessor,
          {
            x0: x0 + uPadding,
            y0: y0 + topPadding,
            width,
            height,
          },
          layouts,
        );
      }),
    )
    .concat(
      treemap(
        nodes.slice(vector.length),
        areaAccessor,
        topPaddingAccessor,
        paddingAccessor,
        vertical
          ? { x0: outerX0, y0: outerY0 + dependentSize, width: outerWidth, height: outerHeight - dependentSize }
          : { x0: outerX0 + dependentSize, y0: outerY0, width: outerWidth - dependentSize, height: outerHeight },
        layouts,
      ),
    );
}
