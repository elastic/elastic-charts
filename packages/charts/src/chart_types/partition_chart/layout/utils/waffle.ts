/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { CHILDREN_KEY, HierarchyOfArrays } from './group_by_rollup';
import { Part } from '../../../../common/text_utils';

// 10 x 10 grid for 100 cells ie. one cell is 1%
const rowCount = 10;
const columnCount = 10;

/** @internal */
export function waffle(
  tree: HierarchyOfArrays,
  totalValue: number,
  {
    x0: outerX0,
    y0: outerY0,
    width: outerWidth,
    height: outerHeight,
  }: { x0: number; y0: number; width: number; height: number },
): Array<Part> {
  const root = tree[0];
  if (!root || !root[1]) return [];

  const size = Math.min(outerWidth, outerHeight);
  const widthOffset = Math.max(0, outerWidth - size) / 2;
  const heightOffset = Math.max(0, outerHeight - size) / 2;
  const rowHeight = size / rowCount;
  const columnWidth = size / columnCount;
  const cellCount = rowCount * columnCount;
  const valuePerCell = totalValue / cellCount;
  let valueSoFar = 0;
  let lastIndex = 0;

  return [
    { node: root, x0: 0, y0: 0, x1: size, y1: size },
    ...root[1][CHILDREN_KEY].flatMap((entry) => {
      const [, { value }] = entry;
      valueSoFar += value;
      const toIndex = Math.round(valueSoFar / valuePerCell);
      const cells = [];
      for (let i = lastIndex; i < toIndex; i++) {
        const columnIndex = i % columnCount;
        const rowIndex = (i - columnIndex) / columnCount;
        const x0 = outerX0 + widthOffset + columnIndex * columnWidth;
        const y0 = outerY0 + heightOffset + rowIndex * rowHeight;
        cells.push({
          node: entry,
          x0,
          y0,
          x1: x0 + columnWidth,
          y1: y0 + rowHeight,
        });
      }
      lastIndex = toIndex;
      return cells;
    }),
  ];
}
