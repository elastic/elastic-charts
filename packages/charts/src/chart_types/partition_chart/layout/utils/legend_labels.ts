/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { LegendItemLabel } from '../../../../state/selectors/get_legend_items_labels';
import { ValueFormatter } from '../../../../utils/common';
import { Layer } from '../../specs';
import {
  CHILDREN_KEY,
  HIERARCHY_ROOT_KEY,
  HierarchyOfArrays,
  PATH_KEY,
  ArrayNode,
  NULL_SMALL_MULTIPLES_KEY,
  AGGREGATE_KEY,
} from './group_by_rollup';

/** @internal */
export function getLegendLabelsAndValue(
  layers: Layer[],
  tree: HierarchyOfArrays,
  legendMaxDepth: number,
  // if used the resulting label will be concatenated with the formatted value, use () => '' to avoid that
  valueFormatter: ValueFormatter,
): LegendItemLabel[] {
  return flatSlicesNames(layers, 0, tree, valueFormatter).filter(({ depth }) => depth <= legendMaxDepth);
}

/** @internal */
function getArrayNodeKey(arrayNode: ArrayNode): string {
  return arrayNode[PATH_KEY].reduce<string>((acc, { value, index }) => {
    if (value === HIERARCHY_ROOT_KEY || value === NULL_SMALL_MULTIPLES_KEY) return acc;
    return `${acc}(${index}):${value}__`;
  }, '__');
}

function flatSlicesNames(
  layers: Layer[],
  depth: number,
  tree: HierarchyOfArrays,
  valueFormatter: ValueFormatter,
  keys: Map<string, string> = new Map(),
  depths: Map<string, number> = new Map(),
): LegendItemLabel[] {
  if (tree.length === 0) {
    return [];
  }

  for (const [key, arrayNode] of tree) {
    // format the key with the layer formatter
    const layer = layers[depth - 1];
    const formatter = layer?.nodeLabel;
    const formattedKey = formatter ? formatter(key) : `${key}`;
    // preventing errors from external formatters
    if (formattedKey && formattedKey !== HIERARCHY_ROOT_KEY) {
      // Node key must be unique for each node in the tree
      const nodeKey = getArrayNodeKey(arrayNode);
      // save only the max depth, so we can compute the the max extension of the legend
      depths.set(nodeKey, depth);

      const formattedValue = valueFormatter(arrayNode[AGGREGATE_KEY]);
      keys.set(nodeKey, `${formattedKey}${formattedValue}`);
    }

    const children = arrayNode[CHILDREN_KEY];
    flatSlicesNames(layers, depth + 1, children, valueFormatter, keys, depths);
  }

  return [...depths.keys()].map((key) => ({
    label: keys.get(key) ?? '',
    depth: depths.get(key) ?? 0,
  }));
}
