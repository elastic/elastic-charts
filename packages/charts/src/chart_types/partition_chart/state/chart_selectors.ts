/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { computeLegendSelector } from './selectors/compute_legend';
import { getChartTypeDescriptionSelector } from './selectors/get_chart_type_description';
import { getPointerCursorSelector } from './selectors/get_cursor_pointer';
import { getDebugStateSelector } from './selectors/get_debug_state';
import { getLegendItemsLabels } from './selectors/get_legend_items_labels';
import { isTooltipVisibleSelector } from './selectors/is_tooltip_visible';
import { createOnElementClickCaller } from './selectors/on_element_click_caller';
import { createOnElementOutCaller } from './selectors/on_element_out_caller';
import { createOnElementOverCaller } from './selectors/on_element_over_caller';
import { getPartitionSpec } from './selectors/partition_spec';
import { getTooltipInfoSelector } from './selectors/tooltip';
import { EMPTY_LEGEND_ITEM_EXTRA_VALUES } from '../../../common/legend';
import type { ChartSelectorsFactory } from '../../../state/chart_selectors';
import type { GlobalChartState } from '../../../state/chart_state';
import { getActivePointerPosition } from '../../../state/selectors/get_active_pointer_position';
import { InitStatus } from '../../../state/selectors/get_internal_is_intialized';

/** @internal */
export const chartSelectorsFactory: ChartSelectorsFactory = () => {
  const onElementClickCaller = createOnElementClickCaller();
  const onElementOverCaller = createOnElementOverCaller();
  const onElementOutCaller = createOnElementOutCaller();

  return {
    isInitialized: (state: GlobalChartState) =>
      getPartitionSpec(state) !== null ? InitStatus.Initialized : InitStatus.SpecNotInitialized,

    isBrushAvailable: () => false,
    isBrushing: () => false,
    isChartEmpty: () => false,

    getLegendItems: computeLegendSelector,
    // order doesn't matter, but it needs to return the highest depth of the label occurrence so enough horizontal width is allocated
    // the label item strings needs to be a concatenation of the label + the extra formatted value if available.
    // this is required to compute the legend automatic width
    getLegendItemsLabels,
    getLegendExtraValues: () => EMPTY_LEGEND_ITEM_EXTRA_VALUES,
    getPointerCursor: getPointerCursorSelector,

    isTooltipVisible: (globalState: GlobalChartState) => ({
      visible: isTooltipVisibleSelector(globalState),
      isExternal: false,
      displayOnly: false,
      isPinnable: true,
    }),
    getTooltipInfo: getTooltipInfoSelector,
    getTooltipAnchor: (state: GlobalChartState) => {
      const position = getActivePointerPosition(state);
      return {
        isRotated: false,
        x: position.x,
        width: 0,
        y: position.y,
        height: 0,
      };
    },

    eventCallbacks: (state: GlobalChartState) => {
      onElementOverCaller(state);
      onElementOutCaller(state);
      onElementClickCaller(state);
    },

    // TODO
    getProjectionContainerArea: () => ({ width: 0, height: 0, top: 0, left: 0 }),

    // TODO
    getMainProjectionArea: () => ({ width: 0, height: 0, top: 0, left: 0 }),

    // TODO
    getBrushArea: () => null,

    getDebugState: getDebugStateSelector,
    getChartTypeDescription: getChartTypeDescriptionSelector,
    getSmallMultiplesDomains: () => ({
      smHDomain: [],
      smVDomain: [],
    }),
    canDisplayChartTitles: () => true,
  };
};
