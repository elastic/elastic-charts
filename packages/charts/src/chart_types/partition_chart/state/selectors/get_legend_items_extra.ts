/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { LegendItemExtraValues } from '../../../../common/legend';
import { SeriesKey } from '../../../../common/series_id';
import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_spec';
import { getExtraValueMap } from '../../layout/viewmodel/hierarchy_of_arrays';
import { getPartitionSpec } from './partition_spec';
import { getTrees } from './tree';

/** @internal */
export const getLegendItemsExtra = createCustomCachedSelector(
  [getPartitionSpec, getSettingsSpecSelector, getTrees],
  (spec, { legendMaxDepth }, trees): Map<SeriesKey, LegendItemExtraValues> => {
    const emptyMap = new Map<SeriesKey, LegendItemExtraValues>();
    return spec && !Number.isNaN(legendMaxDepth) && legendMaxDepth > 0
      ? trees.reduce((result, { tree }) => {
          const treeData = getExtraValueMap(spec.layers, spec.valueFormatter, tree, legendMaxDepth);
          for (const [key, value] of treeData) {
            result.set(key, value);
          }
          return result;
        }, emptyMap)
      : emptyMap;
  },
);
