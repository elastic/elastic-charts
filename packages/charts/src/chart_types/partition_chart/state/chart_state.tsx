/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { RefObject } from 'react';

import { ChartType } from '../..';
import { BackwardRef, GlobalChartState, InternalChartState } from '../../../state/chart_state';
import { getActivePointerPosition } from '../../../state/selectors/get_active_pointer_position';
import { InitStatus } from '../../../state/selectors/get_internal_is_intialized';
import { DebugState } from '../../../state/types';
import { Dimensions } from '../../../utils/dimensions';
import { render } from '../renderer/dom/layered_partition_chart';
import { computeLegendSelector } from './selectors/compute_legend';
import { getChartTypeDescriptionSelector } from './selectors/get_chart_type_description';
import { getPointerCursorSelector } from './selectors/get_cursor_pointer';
import { getDebugStateSelector } from './selectors/get_debug_state';
import { getLegendItemsExtra } from './selectors/get_legend_items_extra';
import { getLegendItemsLabels } from './selectors/get_legend_items_labels';
import { isTooltipVisibleSelector } from './selectors/is_tooltip_visible';
import { createOnElementClickCaller } from './selectors/on_element_click_caller';
import { createOnElementOutCaller } from './selectors/on_element_out_caller';
import { createOnElementOverCaller } from './selectors/on_element_over_caller';
import { getPartitionSpec } from './selectors/partition_spec';
import { getTooltipInfoSelector } from './selectors/tooltip';

/** @internal */
export class PartitionState implements InternalChartState {
  chartType = ChartType.Partition;

  onElementClickCaller: (state: GlobalChartState) => void;

  onElementOverCaller: (state: GlobalChartState) => void;

  onElementOutCaller: (state: GlobalChartState) => void;

  constructor() {
    this.onElementClickCaller = createOnElementClickCaller();
    this.onElementOverCaller = createOnElementOverCaller();
    this.onElementOutCaller = createOnElementOutCaller();
  }

  isInitialized(globalState: GlobalChartState) {
    return getPartitionSpec(globalState) !== null ? InitStatus.Initialized : InitStatus.SpecNotInitialized;
  }

  isBrushAvailable() {
    return false;
  }

  isBrushing() {
    return false;
  }

  isChartEmpty() {
    return false;
  }

  getLegendItemsLabels(globalState: GlobalChartState) {
    // order doesn't matter, but it needs to return the highest depth of the label occurrence so enough horizontal width is allocated
    // the label item strings needs to be a concatenation of the label + the extra formatted value if available.
    // this is required to compute the legend automatic width
    return getLegendItemsLabels(globalState);
  }

  getLegendItems(globalState: GlobalChartState) {
    return computeLegendSelector(globalState);
  }

  getLegendExtraValues(globalState: GlobalChartState) {
    return getLegendItemsExtra(globalState);
  }

  chartRenderer(containerRef: BackwardRef, forwardStageRef: RefObject<HTMLCanvasElement>) {
    return render(containerRef, forwardStageRef);
  }

  getPointerCursor(globalState: GlobalChartState) {
    return getPointerCursorSelector(globalState);
  }

  isTooltipVisible(globalState: GlobalChartState) {
    return {
      visible: isTooltipVisibleSelector(globalState),
      isExternal: false,
      displayOnly: false,
      isPinnable: true,
    };
  }

  getTooltipInfo(globalState: GlobalChartState) {
    return getTooltipInfoSelector(globalState);
  }

  getTooltipAnchor(state: GlobalChartState) {
    const position = getActivePointerPosition(state);
    return {
      isRotated: false,
      x: position.x,
      width: 0,
      y: position.y,
      height: 0,
    };
  }

  eventCallbacks(globalState: GlobalChartState) {
    this.onElementOverCaller(globalState);
    this.onElementOutCaller(globalState);
    this.onElementClickCaller(globalState);
  }

  // TODO
  getProjectionContainerArea(): Dimensions {
    return { width: 0, height: 0, top: 0, left: 0 };
  }

  // TODO
  getMainProjectionArea(): Dimensions {
    return { width: 0, height: 0, top: 0, left: 0 };
  }

  // TODO
  getBrushArea(): Dimensions | null {
    return null;
  }

  getDebugState(state: GlobalChartState): DebugState {
    return getDebugStateSelector(state);
  }

  getChartTypeDescription(state: GlobalChartState): string {
    return getChartTypeDescriptionSelector(state);
  }
}
