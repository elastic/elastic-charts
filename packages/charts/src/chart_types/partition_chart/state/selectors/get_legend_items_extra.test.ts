/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { Store } from 'redux';

import { getLegendItemsExtra } from './get_legend_items_extra';
import { MockSeriesSpec, MockGlobalSpec } from '../../../../mocks/specs';
import { MockStore } from '../../../../mocks/store';
import { GlobalChartState } from '../../../../state/chart_state';
import { PrimitiveValue } from '../../layout/utils/group_by_rollup';

describe('Partition - Legend item extra values', () => {
  type TestDatum = [string, string, string, number];
  const spec = MockSeriesSpec.sunburst({
    data: [
      ['aaa', 'aa', '1', 1],
      ['aaa', 'aa', '1', 2],
      ['aaa', 'aa', '3', 1],
      ['aaa', 'bb', '4', 1],
      ['aaa', 'bb', '5', 1],
      ['aaa', 'bb', '6', 1],
      ['bbb', 'aa', '7', 1],
      ['bbb', 'aa', '8', 1],
      ['bbb', 'bb', '9', 1],
      ['bbb', 'bb', '10', 1],
      ['bbb', 'cc', '11', 1],
      ['bbb', 'cc', '12', 1],
    ],
    valueAccessor: (d: TestDatum) => d[3],
    layers: [
      {
        groupByRollup: (datum: TestDatum) => datum[0],
        nodeLabel: (d: PrimitiveValue) => String(d),
      },
      {
        groupByRollup: (datum: TestDatum) => datum[1],
        nodeLabel: (d: PrimitiveValue) => String(d),
      },
      {
        groupByRollup: (datum: TestDatum) => datum[2],
        nodeLabel: (d: PrimitiveValue) => String(d),
      },
    ],
  });
  let store: Store<GlobalChartState>;

  beforeEach(() => {
    store = MockStore.default();
  });

  it('should return all extra values in nested legend', () => {
    MockStore.addSpecs([spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual([
      '0__0',
      '0__0__0',
      '0__0__0__0',
      '0__0__0__0__0',
      '0__0__0__0__1',
      '0__0__0__1',
      '0__0__0__1__0',
      '0__0__0__1__1',
      '0__0__0__1__2',
      '0__0__1',
      '0__0__1__0',
      '0__0__1__0__0',
      '0__0__1__0__1',
      '0__0__1__1',
      '0__0__1__1__0',
      '0__0__1__1__1',
      '0__0__1__2',
      '0__0__1__2__0',
      '0__0__1__2__1',
    ]);
    expect(extraValues.values()).toMatchSnapshot();
  });

  it('should return extra values in nested legend within max depth of 1', () => {
    const settings = MockGlobalSpec.settings({ legendMaxDepth: 1 });
    MockStore.addSpecs([settings, spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual(['0__0', '0__0__0', '0__0__1']);
    expect(extraValues.values()).toMatchSnapshot();
  });

  it('should return extra values in nested legend within max depth of 2', () => {
    const settings = MockGlobalSpec.settings({ legendMaxDepth: 2 });
    MockStore.addSpecs([settings, spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual([
      '0__0',
      '0__0__0',
      '0__0__0__0',
      '0__0__0__1',
      '0__0__1',
      '0__0__1__0',
      '0__0__1__1',
      '0__0__1__2',
    ]);
    expect(extraValues.values()).toMatchSnapshot();
  });

  it('filters all extraValues if depth is 0', () => {
    const settings = MockGlobalSpec.settings({ legendMaxDepth: 0 });
    MockStore.addSpecs([settings, spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual([]);
  });

  it('filters all extraValues if depth is NaN', () => {
    const settings = MockGlobalSpec.settings({ legendMaxDepth: NaN });
    MockStore.addSpecs([settings, spec], store);

    const extraValues = getLegendItemsExtra(store.getState());
    expect([...extraValues.keys()]).toEqual([]);
  });
});
