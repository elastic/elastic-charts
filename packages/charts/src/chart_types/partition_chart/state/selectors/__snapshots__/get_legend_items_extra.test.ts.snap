// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Partition - Legend item extra values should return all extra values in nested legend 1`] = `{}`;

exports[`Partition - Legend item extra values should return extra values in nested legend within max depth of 1 1`] = `{}`;

exports[`Partition - Legend item extra values should return extra values in nested legend within max depth of 2 1`] = `{}`;
